framework/filters/AccessRule.php

```diff
@@ -136,19 +142,30 @@ protected function matchController($controller)
      */
     protected function matchRole($user)
     {
-        if (empty($this->roles)) {
+        if (empty($this->roles) && empty($this->permissions)) {
             return true;
         }
-        foreach ($this->roles as $role) {
-            if ($role === '?') {
+
+        $privileges = [];
+
+        if($this->roles) {
+            $privileges = array_merge($privileges, $this->roles);
```

Comment:
You can do `$roles = ($this->roles === null) ? [] : $this->roles` and then merge permissions if they are not `null`:


```php
$roles = empty($this->roles) ? [] : $this->roles;
if (!empty($this->permissions)) {
   $roles = array_merge($roles, $this->permissions);
}
```